File: demo-web-scala/src/main/scala/com/example/demo/security/CustomUserDetailsService.scala
package com.example.demo.security

import com.example.demo.repository.UserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.userdetails.{UserDetails, UserDetailsService, UsernameNotFoundException}
import org.springframework.transaction.annotation.Transactional

class CustomUserDetailsService  extends UserDetailsService {
  @Autowired var userRepository: UserRepository = null
  @Transactional(readOnly = true)
  @throws[UsernameNotFoundException]
  override def loadUserByUsername(username: String): UserDetails = {
    val user = userRepository.findByUsername(username)
    new CustomUserDetails(user)
  }
}

File: demo-web-java/src/main/java/com/example/demo/security/CustomUserDetailsService.java
package com.example.demo.security;

import com.example.demo.model.User;
import com.example.demo.repository.UserRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.transaction.annotation.Transactional;

import javax.servlet.http.HttpSession;

@Slf4j
public class CustomUserDetailsService implements UserDetailsService {
  @Autowired
  private UserRepository userRepository;

  @Autowired
  HttpSession httpSession;

  @Override
  @Transactional(readOnly = true)
  public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
    log.debug("{}", httpSession.getId());
    User user = userRepository.findByUsername(username);
    return new CustomUserDetails(user);
  }
}
